File: ralib/src/main/java/com/app/ralib/extractors/BasicSevenZipExtractor.java
package com.app.ralib.extractors;

import android.util.Log;

import androidx.annotation.Nullable;

import net.sf.sevenzipjbinding.ExtractAskMode;
import net.sf.sevenzipjbinding.ExtractOperationResult;
import net.sf.sevenzipjbinding.IArchiveExtractCallback;
import net.sf.sevenzipjbinding.IInArchive;
import net.sf.sevenzipjbinding.ISequentialOutStream;
import net.sf.sevenzipjbinding.PropID;
import net.sf.sevenzipjbinding.SevenZip;
import net.sf.sevenzipjbinding.SevenZipException;
import net.sf.sevenzipjbinding.impl.RandomAccessFileInStream;

import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;

public class BasicSevenZipExtractor implements ExtractorCollection.IExtractor {
    private static final String TAG = "BasicSevenZipExtractor";
    private Path sourcePath;
    private Path sourceExtractionPrefix;
    private Path destinationPath;
    private ExtractorCollection.ExtractionListener extractionListener;
    private HashMap<String, Object> state;

    public BasicSevenZipExtractor(Path sourcePath, Path destinationPath) {
        this.setSourcePath(sourcePath);
        this.setDestinationPath(destinationPath);
        this.setSourceExtractionPrefix(Paths.get(""));
    }

    public BasicSevenZipExtractor(Path sourcePath, Path destinationPath, ExtractorCollection.ExtractionListener listener) {
        this.setSourcePath(sourcePath);
        this.setDestinationPath(destinationPath);
        this.setExtractionListener(listener);
        this.setSourceExtractionPrefix(Paths.get(""));
    }

    public BasicSevenZipExtractor(Path sourcePath, Path sourceExtractionPrefix, Path destinationPath, ExtractorCollection.ExtractionListener listener) {
        this.setSourcePath(sourcePath);
        this.setDestinationPath(destinationPath);
        this.setExtractionListener(listener);
        this.setSourceExtractionPrefix(sourceExtractionPrefix);
    }

    @Override
    public void setSourcePath(Path sourcePath) {
        this.sourcePath = sourcePath;
    }

    public void setSourceExtractionPrefix(Path sourceExtractionPrefix) {
        this.sourceExtractionPrefix = sourceExtractionPrefix;
    }

    @Override
    public void setDestinationPath(Path destinationPath) {
        this.destinationPath = destinationPath;
    }

    @Override
    public void setExtractionListener(ExtractorCollection.ExtractionListener listener) {
        this.extractionListener = listener;
    }

    @Override
    public void setState(HashMap<String, Object> state) {
        this.state = state;
    }

    @Override
    public HashMap<String, Object> getState() {
        return this.state;
    }

    @Override
    public boolean extract() {
        try {
            if (!Files.exists(destinationPath)) {
                Files.createDirectories(destinationPath);
            }

            try (RandomAccessFile raf = new RandomAccessFile(sourcePath.toString(), "r");
                 RandomAccessFileInStream is = new RandomAccessFileInStream(raf);
                 IInArchive archive = SevenZip.openInArchive(null, is)) {

                int totalItems = archive.getNumberOfItems();
                Log.d(TAG, "Archive contains " + totalItems + " items");

                // 提取所有文件
                archive.extract(null, false, new ArchiveExtractCallback(archive));
            }

            Log.d(TAG, "SevenZip extraction completed successfully");
            if (extractionListener != null) {
                extractionListener.onProgress("解压完成", 1.0f, state);
                extractionListener.onComplete("解压完成", state);
            }
            return true;
        }
        catch (Exception ex) {
            if (extractionListener != null) {
                extractionListener.onError("基础7z解压器解压失败", ex, state);
            }
        }
        return false;
    }

    /*
     * SevenZipJBinding 提取回调实现
     */
    private class ArchiveExtractCallback implements IArchiveExtractCallback {
        private final IInArchive archive;
        @Nullable
        private SequentialFileOutputStream outputStream = null;
        @Nullable
        private Path currentProcessingFilePath = null;
        private long totalBytes = 0;
        private long totalBytesExtracted = 0;

        public ArchiveExtractCallback(IInArchive archive) {
            this.archive = archive;
        }


        @Override
        public ISequentialOutStream getStream(int index, ExtractAskMode extractAskMode) throws SevenZipException {
            try {
                closeOutputStream();

                // 获取文件信息
                Path filePath = Paths.get(archive.getStringProperty(index, PropID.PATH));
                boolean isFolder = (Boolean) archive.getProperty(index, PropID.IS_FOLDER);

                Log.d(TAG, "Processing item: " + filePath + " (isFolder: " + isFolder + ")");

                // 跳过非指定前缀的文件
                Path relativeFilePath = sourceExtractionPrefix.relativize(filePath).normalize();
                if (relativeFilePath.toString().startsWith("..")) {
                    return null;
                }

                // 计算目标文件路径并防止路径遍历攻击
                Path targetFilePath = destinationPath.resolve(relativeFilePath).normalize();
                if (destinationPath.relativize(targetFilePath).toString().startsWith("..")) {
                    throw new SevenZipException("Attempting to write outside of destination directory: " + targetFilePath);
                }

                // 对于文件夹只创建文件夹
                if (isFolder) {
                    Files.createDirectories(targetFilePath);
                    return null;
                }

                // 创建文件的父目录
                currentProcessingFilePath = targetFilePath;
                Path targetFileParentPath = targetFilePath.normalize().getParent();
                if (!Files.exists(targetFileParentPath)) {
                    Files.createDirectories(targetFileParentPath);
                }

                float progress = (totalBytes > 0) ? (float) totalBytesExtracted / totalBytes : 0f;
                extractionListener.onProgress(String.format("正在解压: %s", filePath.toString()), progress, state);

                // 返回输出流
                return new SequentialFileOutputStream(targetFilePath);
            } catch (Exception e) {
                throw new SevenZipException("Error getting stream for index " + index, e);
            }
        }

        @Override
        public void prepareOperation(ExtractAskMode extractAskMode) throws SevenZipException {

        }

        @Override
        public void setOperationResult(ExtractOperationResult extractOperationResult) throws SevenZipException {
            closeOutputStream();
        }

        @Override
        public void setTotal(long total) throws SevenZipException {
            totalBytes = total;
        }

        @Override
        public void setCompleted(long complete) throws SevenZipException {
            totalBytesExtracted = complete;
        }

        private void closeOutputStream() throws SevenZipException {
            if (outputStream != null) {
                try {
                    outputStream.close();
                    outputStream = null;
                } catch (IOException e) {
                    throw new SevenZipException("Error closing file: "
                            + currentProcessingFilePath);
                }
            }
        }
    }

    /*
     * SevenZipJBinding 输出流实现
     */
    private static class SequentialFileOutputStream implements ISequentialOutStream {
        private final FileOutputStream fileStream;

        public SequentialFileOutputStream(Path targetFilePath) throws FileNotFoundException {
            this.fileStream = new FileOutputStream(targetFilePath.toFile());
        }

        @Override
        public int write(byte[] data) throws SevenZipException {
            try {
                fileStream.write(data);
                return data.length;
            } catch (IOException e) {
                throw new SevenZipException("Error writing to output stream", e);
            }
        }

        public void close() throws IOException {
            fileStream.close();
        }
    }
}

File: app/src/main/java/com/app/ralaunch/core/platform/install/extractors/BasicSevenZipExtractor.kt
package com.app.ralaunch.core.platform.install.extractors

import android.util.Log
import com.app.ralaunch.R
import com.app.ralaunch.RaLaunchApp
import com.app.ralaunch.core.platform.runtime.RuntimeLibraryLoader
import net.sf.sevenzipjbinding.*
import net.sf.sevenzipjbinding.impl.RandomAccessFileInStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.RandomAccessFile
import java.util.zip.ZipInputStream

class BasicSevenZipExtractor : ExtractorCollection.IExtractor {

    companion object {
        private const val TAG = "BasicSevenZipExtractor"

        @Volatile
        private var libraryLoaded = false

        @Synchronized
        fun ensureLibraryLoaded(): Boolean {
            if (libraryLoaded) return true

            try {
                System.loadLibrary("7-Zip-JBinding")
                libraryLoaded = true
                Log.i(TAG, "7-Zip native library loaded successfully via System.loadLibrary")
            } catch (e: UnsatisfiedLinkError) {
                Log.e(TAG, "Failed to load 7-Zip native library: ${e.message}")
                try {
                    val context = RaLaunchApp.getInstance()
                    libraryLoaded = RuntimeLibraryLoader.load7Zip(context)
                    if (libraryLoaded) {
                        Log.i(TAG, "7-Zip native library loaded from runtime_libs")
                    }
                } catch (e2: Exception) {
                    Log.e(TAG, "Cannot load 7-Zip library from runtime_libs: ${e2.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Cannot load 7-Zip library: ${e.message}")
            }
            return libraryLoaded
        }
    }

    private lateinit var sourceFile: File
    private var sourceExtractionPrefix: File = File("")
    private lateinit var destinationFile: File
    private var extractionListener: ExtractorCollection.ExtractionListener? = null
    override var state: HashMap<String, Any?> = hashMapOf()

    constructor(sourceFile: File, destinationFile: File) {
        setSourcePath(sourceFile)
        setDestinationPath(destinationFile)
    }

    constructor(sourceFile: File, destinationFile: File, listener: ExtractorCollection.ExtractionListener?) {
        setSourcePath(sourceFile)
        setDestinationPath(destinationFile)
        setExtractionListener(listener)
    }

    constructor(
        sourceFile: File,
        sourceExtractionPrefix: File,
        destinationFile: File,
        listener: ExtractorCollection.ExtractionListener?
    ) {
        setSourcePath(sourceFile)
        setDestinationPath(destinationFile)
        setExtractionListener(listener)
        setSourceExtractionPrefix(sourceExtractionPrefix)
    }

    override fun setSourcePath(sourcePath: File) {
        this.sourceFile = sourcePath
    }

    fun setSourceExtractionPrefix(sourceExtractionPrefix: File) {
        this.sourceExtractionPrefix = sourceExtractionPrefix
    }

    override fun setDestinationPath(destinationPath: File) {
        this.destinationFile = destinationPath
    }

    override fun setExtractionListener(listener: ExtractorCollection.ExtractionListener?) {
        this.extractionListener = listener
    }

    override fun extract(): Boolean {
        if (!destinationFile.exists()) {
            destinationFile.mkdirs()
        }

        // ... Try 7-Zip engine first ...
        if (ensureLibraryLoaded()) {
            try {
                RandomAccessFile(sourceFile, "r").use { raf ->
                    RandomAccessFileInStream(raf).use { inStream ->
                        SevenZip.openInArchive(null, inStream).use { archive ->
                            val totalItems = archive.numberOfItems
                            Log.d(TAG, "Archive contains $totalItems items")
                            archive.extract(null, false, ArchiveExtractCallback(archive))
                        }
                    }
                }

                Log.d(TAG, "SevenZip extraction completed successfully")
                extractionListener?.apply {
                    val completeMessage = RaLaunchApp.getInstance().getString(R.string.extract_complete)
                    onProgress(completeMessage, 1.0f, state)
                    onComplete(completeMessage, state)
                }
                return true
            } catch (ex: Exception) {
                Log.e(TAG, "7-Zip extraction failed: ${ex.message}", ex)
                // ... DO NOT RETURN FALSE YET, FALLTHROUGH TO LIFEBOAT ...
            }
        } else {
            Log.e(TAG, "7-Zip library is not loaded. Proceeding to fallback.")
        }

        // =====================================================================
        // ... THE LIFEBOAT: Pure Java Zip Fallback for older Android devices ...
        // =====================================================================
        if (sourceFile.name.lowercase().endsWith(".zip")) {
            Log.w(TAG, "Using Standard Java Zip Fallback for: ${sourceFile.name}")
            return fallbackExtractZip()
        }

        // ... If it's not a zip and 7-zip failed, then we truly failed ...
        extractionListener?.onError(
            RaLaunchApp.getInstance().getString(R.string.extract_7zip_failed),
            RuntimeException("Both 7-Zip and Fallback failed"),
            state
        )
        return false
    }

    // =====================================================================
    // ... PURE JAVA ZIP EXTRACTOR (With Strict Zip Slip Protection) ...
    // =====================================================================
    private fun fallbackExtractZip(): Boolean {
        return try {
            FileInputStream(sourceFile).use { fis ->
                ZipInputStream(fis).use { zis ->
                    var bytesExtracted = 0L
                    while (true) {
                        val entry = zis.nextEntry ?: break
                        
                        val filePath = entry.name
                        val prefix = sourceExtractionPrefix.path
                        
                        val relativeFilePath = if (prefix.isEmpty() || prefix == ".") {
                            filePath
                        } else if (filePath.startsWith(prefix)) {
                            filePath.substring(prefix.length).trimStart('/', '\\')
                        } else {
                            zis.closeEntry()
                            continue
                        }

                        // ... Calculate target file and enforce STRICT Zip Slip protection ...
                        val targetFile = File(destinationFile, relativeFilePath).canonicalFile
                        val destCanonicalPath = destinationFile.canonicalPath
                        
                        // ... Append separator to ensure strict directory boundary ...
                        val safeDestPath = if (destCanonicalPath.endsWith(File.separator)) {
                            destCanonicalPath
                        } else {
                            "$destCanonicalPath${File.separator}"
                        }

                        if (!targetFile.canonicalPath.startsWith(safeDestPath)) {
                            throw IOException("Zip Slip / Path traversal detected: $targetFile")
                        }

                        if (entry.isDirectory) {
                            targetFile.mkdirs()
                        } else {
                            // ... Create parent directories ...
                            targetFile.parentFile?.mkdirs()

                            extractionListener?.onProgress(
                                RaLaunchApp.getInstance().getString(R.string.extract_in_progress, filePath),
                                0.5f, 
                                state
                            )

                            FileOutputStream(targetFile).use { fos ->
                                val buffer = ByteArray(8192)
                                while (true) {
                                    val read = zis.read(buffer)
                                    if (read <= 0) break
                                    fos.write(buffer, 0, read)
                                    bytesExtracted += read
                                }
                            }
                        }
                        zis.closeEntry()
                    }
                }
            }

            Log.d(TAG, "Fallback extraction completed successfully")
            extractionListener?.apply {
                val completeMessage = RaLaunchApp.getInstance().getString(R.string.extract_complete)
                onProgress(completeMessage, 1.0f, state)
                onComplete(completeMessage, state)
            }
            true
        } catch (ex: Exception) {
            Log.e(TAG, "Fallback ZIP extraction also failed", ex)
            extractionListener?.onError("Zip Fallback Extraction Failed", ex, state)
            false
        }
    }

    // =====================================================================
    // ... 7-ZIP CALLBACK HANDLER (With Strict Zip Slip Protection) ...
    // =====================================================================
    private inner class ArchiveExtractCallback(
        private val archive: IInArchive
    ) : IArchiveExtractCallback {

        private var outputStream: SequentialFileOutputStream? = null
        private var currentProcessingFile: File? = null
        private var totalBytes: Long = 0
        private var totalBytesExtracted: Long = 0

        @Throws(SevenZipException::class)
        override fun getStream(index: Int, extractAskMode: ExtractAskMode): ISequentialOutStream? {
            try {
                closeOutputStream()

                val filePath = archive.getStringProperty(index, PropID.PATH) ?: ""
                val isFolder = archive.getProperty(index, PropID.IS_FOLDER) as? Boolean ?: false

                val prefix = sourceExtractionPrefix.path
                val relativeFilePath = if (prefix.isEmpty() || prefix == ".") {
                    filePath
                } else if (filePath.startsWith(prefix)) {
                    filePath.substring(prefix.length).trimStart('/', '\\')
                } else {
                    return null
                }

                // ... Calculate target file and enforce STRICT Zip Slip protection ...
                val targetFile = File(destinationFile, relativeFilePath).canonicalFile
                val destCanonicalPath = destinationFile.canonicalPath
                
                // ... Append separator to ensure strict directory boundary ...
                val safeDestPath = if (destCanonicalPath.endsWith(File.separator)) {
                    destCanonicalPath
                } else {
                    "$destCanonicalPath${File.separator}"
                }

                if (!targetFile.canonicalPath.startsWith(safeDestPath)) {
                    throw SevenZipException("Zip Slip / Path traversal detected: $targetFile")
                }

                if (isFolder) {
                    targetFile.mkdirs()
                    return null
                }

                currentProcessingFile = targetFile
                targetFile.parentFile?.mkdirs()

                val progress = if (totalBytes > 0) totalBytesExtracted.toFloat() / totalBytes else 0f
                extractionListener?.onProgress(
                    RaLaunchApp.getInstance().getString(R.string.extract_in_progress, filePath),
                    progress,
                    state
                )

                outputStream = SequentialFileOutputStream(targetFile)
                return outputStream
            } catch (e: Exception) {
                throw SevenZipException("Error getting stream for index $index", e)
            }
        }

        @Throws(SevenZipException::class)
        override fun prepareOperation(extractAskMode: ExtractAskMode) {}

        @Throws(SevenZipException::class)
        override fun setOperationResult(extractOperationResult: ExtractOperationResult) {
            closeOutputStream()
        }

        @Throws(SevenZipException::class)
        override fun setTotal(total: Long) {
            totalBytes = total
        }

        @Throws(SevenZipException::class)
        override fun setCompleted(complete: Long) {
            totalBytesExtracted = complete
        }

        @Throws(SevenZipException::class)
        private fun closeOutputStream() {
            outputStream?.let {
                try {
                    it.close()
                    outputStream = null
                } catch (e: IOException) {
                    throw SevenZipException("Error closing file: $currentProcessingFile")
                }
            }
        }
    }

    private class SequentialFileOutputStream(targetFile: File) : ISequentialOutStream {
        private val fileStream = FileOutputStream(targetFile)

        @Throws(SevenZipException::class)
        override fun write(data: ByteArray): Int {
            return try {
                fileStream.write(data)
                data.size
            } catch (e: IOException) {
                throw SevenZipException("Error writing to output stream", e)
            }
        }

        @Throws(IOException::class)
        fun close() {
            fileStream.close()
        }
    }
}
